### PR TITLE
Feature/session and session routes

### PR DESCRIPTION
**Pull request by Luke-Clements**
# Description

Session Creation, Validation, Updating, and Deletion fully implemented and tested, except the controller.

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Units tests included.

### CODE REVIEW

File: Backend/DataAccessLayer/Repositories/SessionRepository.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+using System.Data.Entity;
+using System.Linq;
+
+namespace DataAccessLayer.Repositories
+{
+    public class SessionRepository
+    {
+        //returns null if no valid session is found in the sessions table, otherwise
+        //  returns the current session
+        public Session GetSession(DatabaseContext _db, string token, Guid userId)
```
Comment (ChrisMeyer7088) [inline Backend/DataAccessLayer/Repositories/SessionRepository.cs]: When getSession() is called the userID will not be accessible.

Comment (ChrisMeyer7088) [reply Backend/DataAccessLayer/Repositories/SessionRepository.cs]: also is token referring to ID of session? If so it is a Universally Unique Identifier so no need to use the UserId as well

File: Backend/DataAccessLayer/Repositories/SessionRepository.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        
+        public Session CreateSession(DatabaseContext _db, Session session)
+        {
+            _db.Entry(session).State = EntityState.Added;
+            return session;
+        }
+
+        public Session ValidateSession(DatabaseContext _db, string token, Guid userId)
+        {
+            Session session = GetSession(_db, token, userId);
```
Comment (ChrisMeyer7088) [inline Backend/DataAccessLayer/Repositories/SessionRepository.cs]: Need to remove userID from this call

File: Backend/ServiceLayer/Services/SessionService.cs [hunk 3]
```diff
@@ -1,35 +1,42 @@
...
+        }
+
+        public Session DeleteSession(DatabaseContext _db, string token, Guid userId)
+        {
+            return _SessionRepo.DeleteSession(_db, token, userId);
+        }
+
+        public Session GetSession(DatabaseContext _db, string token, Guid userId)
+        {
+            return _SessionRepo.GetSession(_db, token, userId);
```
Comment (ChrisMeyer7088) [inline Backend/ServiceLayer/Services/SessionService.cs]: UserId will not be passed in

File: Backend/UnitTesting/SessionServiceUT.cs [hunk 4]
```diff
@@ -9,37 +10,248 @@ namespace UnitTesting
...
+            newUser = tu.CreateUserObject();
+            newSession = tu.CreateSessionObject(newUser);
+            var expectedResult = newSession;
+
+            // ACT
+            using (_db = tu.CreateDataBaseContext())
+            {
+                newSession = ss.CreateSession(_db, newSession);
+                _db.SaveChanges();
+                var result = ss.GetSession(_db, newSession.Token, newUser.Id);
```
Comment (ChrisMeyer7088) [inline Backend/UnitTesting/SessionServiceUT.cs]: Here as well

File: Backend/UnitTesting/SessionServiceUT.cs [hunk 5]
```diff
@@ -9,37 +10,248 @@ namespace UnitTesting
...
+        {
+            // Arrange
+            Guid nonExistingSession = Guid.NewGuid();
+            string nonExistingToken = Guid.NewGuid().ToString();
+            Session expectedResult = null;
+
+            // Act
+            using (_db = tu.CreateDataBaseContext())
+            {
+                var result = ss.GetSession(_db, nonExistingToken, nonExistingSession);
```
Comment (ChrisMeyer7088) [inline Backend/UnitTesting/SessionServiceUT.cs]: Here as well, nonExistingSession is actually a userID param, needs to be removed

**Review by ChrisMeyer7088 — CHANGES_REQUESTED** [review]

File: Backend/DataAccessLayer/Repositories/SessionRepository.cs (hunk 1 above)
Comment (Luke-Clements) [reply Backend/DataAccessLayer/Repositories/SessionRepository.cs]: the token is the actual token the user was given, not the Guid of the token in the session database

**Review by ChrisMeyer7088 — APPROVED** [review]

**Pull request merged**